core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala

```diff
@@ -166,9 +172,12 @@ private[spark] class MetricsSystem private (
   }
 
   def removeSource(source: Source) {
-    sources -= source
-    val regName = buildRegistryName(source)
-    registry.removeMatching((name: String, _: Metric) => name.startsWith(regName))
+    val index = sources.indexOf(source)
+    if (index != -1) {
+      sources.remove(index)
+      val regNames = getRegisteredNames(source)
+      registry.removeMatching((name: String, _: Metric) => regNames.contains(name))
```

Comment:
Is it really necessary to iterate over all the metrics in the registry to remove the necessary metrics?
Isn't it more efficient just to remove the metrics from the registry by their names using `MetricRegistry.remove(String)`?